extensions/GitExtension/FileExplorerGitIntegration/Models/RepositoryWrapper.cs

```diff
@@ -0,0 +1,198 @@
+{
+    private readonly Repository _repo;
+    private readonly ReaderWriterLockSlim _repoLock = new();
+
+    private readonly string _workingDirectory;
+
+    private Commit? _head;
+    private CommitLogCache? _commits;
+
+    private bool _disposedValue;
+
+    public RepositoryWrapper(string rootFolder)
+    {
+        _repo = new Repository(rootFolder);
+        _workingDirectory = _repo.Info.WorkingDirectory;
+    }
+
+    public IEnumerable<Commit> GetCommits()
+    {
+        // Fast path: if we have an up-to-date commit log, return that
+        if (_head != null && _commits != null)
+        {
+            _repoLock.EnterReadLock();
+            try
+            {
+                if (_repo.Head.Tip == _head)
+                {
+                    return _commits;
+                }
+            }
+            finally
+            {
+                _repoLock?.ExitReadLock();
+            }
+        }
+
+        // Either the commit log hasn't been created yet, or it's out of date
+        _repoLock.EnterWriteLock();
+        try
+        {
+            if (_head == null || _commits == null || _repo.Head.Tip != _head)
+            {
+                _commits = new CommitLogCache(_repo);
+                _head = _repo.Head.Tip;
+            }
+        }
+        finally
+        {
+            _repoLock.ExitWriteLock();
+        }
+
+        return _commits;
+    }
+
+    public string GetRepoStatus()
+    {
+        RepositoryStatus repositoryStatus;
+        string branchName;
+        var branchStatus = string.Empty;
+        try
+        {
+            _repoLock.EnterWriteLock();
+            repositoryStatus = _repo.RetrieveStatus();
+            branchName = _repo.Info.IsHeadDetached ?
+                "Detached: " + _repo.Head.Tip.Sha[..7] :
+                "Branch: " + _repo.Head.FriendlyName;
+            if (_repo.Head.IsTracking)
+            {
+                var behind = _repo.Head.TrackingDetails.BehindBy;
+                var ahead = _repo.Head.TrackingDetails.AheadBy;
+                if (behind == 0 && ahead == 0)
+                {
+                    branchStatus = " ≡";
+                }
+                else if (behind > 0 && ahead > 0)
+                {
+                    branchStatus = " ↓" + behind + " ↑" + ahead;
+                }
+                else if (behind > 0)
+                {
+                    branchStatus = " ↓" + behind;
+                }
+                else if (ahead > 0)
+                {
+                    branchStatus = " ↑" + ahead;
+                }
+            }
+        }
+        finally
+        {
+            _repoLock.ExitWriteLock();
+        }
+
+        var fileStatus = $"| +{repositoryStatus.Added.Count()} ~{repositoryStatus.Staged.Count()} -{repositoryStatus.Removed.Count()} | +{repositoryStatus.Untracked.Count()} ~{repositoryStatus.Modified.Count()} -{repositoryStatus.Missing.Count()}";
+        var conflicted = 0;

```

Comment:
NIT: `conflicted = repositoryStatus.Where(x => x.State.HasFlag(FileStatus.Conflicted).Count();`

Reply:
Might need a `.ToList()` to get the count.